### DISCUSSION TITLE
What is the purpose of running PBFT inside enclave?

### DISCUSSION THREAD

**Discussion by MingyuanGao** [microsoft/CCF#2479] [General] [answered]
Hello, CCF Team, thank you for this excellent open-source project. From the build system, PBFT code is linked to the enclave part. Is it correct to say that _PBFT totally runs inside the enclave_? If this is the case, _what is the purpose of running PBFT inside the enclave_? Thanks in advance.

**Comment by ashamis** [comment, marked as answer]
Hi Mingyuan, we run the consensus protocol inside of an enclave as a way of hardening the protocol as its harder to attack code running inside an enclave as compared to code that is not.

**Reply by MingyuanGao** [reply]
Thanks for your explanation. I am also curious about the performance overhead of running PBFT inside the enclave. Compared with running PBFT outside the enclave, have you measured how much performance overhead is incurred? Will the ECalls and OCalls incur a high overhead?

**Reply by ashamis** [reply]
We use ring buffers for communication so there are no ECalls and OCalls. We found in a older version of the code that overhead affected throughput by 16%. This was inline with other published results.

Meni Orenbach, Pavel Lifshits, Marina Minkin, and Mark Silberstein. Eleos: Exitless os services for sgx enclaves. In Proceedings of the Twelfth European Conference on Computer Systems, pages 238–253, 2017.

**Reply by MingyuanGao** [reply]
@ashamis Thank you so much for your answer.

**Reply by ashamis** [reply]
Just adding a bit to this discussion, the current BFT implementation in CCF is not PBFT it is experimental so it is subject to change and not optimized.

**Reply by MingyuanGao** [reply]
Thanks for this important clarification. We are a research team doing research on consensus protocols. I have been using CCF from ccf-0.13.2, which has a folder "pbft" under src/consensus. From ccf-0.13.3, I saw raft and pbft were merged. Since we need to test our idea using both TEE+Raft and TEE+PBFT, could you help to point out which version has the working PBFT support?  If the current BFT implementation is based on PBFT, it would be appreciated if you could list down what changes CCF has made. Thanks a lot.

**Reply by ashamis** [reply]
Thank you for describing some of what you are trying to do. We are currently in the process of getting this work published and as such we cannot provide a list of changes that we made. There is no version of CCF that runs PBFT without our protocol changes, due to these changes we are not calling it PBFT. If you would like we can take this discussion offline my email is alexsha@microsoft.com 

I would also like to add that we are not running vanilla RAFT and have modified the protocol also.

**Reply by MingyuanGao** [reply]
Thank you so much.

**Reply by Parsifal0** [reply]
> Hi Mingyuan, we run the consensus protocol inside of an enclave as a way of hardening the protocol as its harder to attack code running inside an enclave as compared to code that is not.

Hello, Ashamis. Does moving consensus, i.e., the replication protocols, outside enclaves break the confidentiality/security of CCF? Since the local ledger storage is already encrytped in the disk, it seems that placing consensus in the host side is fine. Thanks for your attention.

**Reply by ashamis** [reply]
@Parsifal0 It is possible to move the Byzantine consensus outside the enclave and still maintain similar security properties. However, this would required a well thought-out change as an naive solution will probably results in some security and confidentiality issues.

**Reply by Parsifal0** [reply]
Thanks! It does make sense. Putting consensus inside enclaves or not is just an implementation choice, which doesn't impact the threat model of CCF.  If one tends to keep consensus outside enclaves, he may need extra steps (e.g., signature verification) to communicate with the consensus.

**Reply by ashamis** [reply]
Its part of a defense in depth strategy
